lib/Crypto/Hash/cSHAKE128.py

```diff
@@ -28,39 +28,100 @@
 # POSSIBILITY OF SUCH DAMAGE.
 # ===================================================================
 
-from Crypto.Util.py3compat import bord
+from Crypto.Util.py3compat import bchr
 
-from Crypto.Util._raw_api import (load_pycryptodome_raw_lib,
-                                  VoidPointer, SmartPointer,
+from Crypto.Util._raw_api import (VoidPointer, SmartPointer,
                                   create_string_buffer,
                                   get_raw_buffer, c_size_t,
                                   c_uint8_ptr)
 
 from Crypto.Hash.keccak import _raw_keccak_lib
 
-class SHAKE128_XOF(object):
-    """A SHAKE128 hash object.
+
+def left_encode(x: int):
+    """Left encode function as defined in NIST SP 800-185"""
+
+    assert (x < (1 << 2040) and x >= 0)
+
+    # Get number of bytes needed to represent this integer.
+    num = 1 if x == 0 else (x.bit_length() + 7) // 8
+    # Chop off least significant byte of integer.
+    enc = bchr(num) + bchr(x & 0xFF)
+    x >>= 8
+    # Chop off remaining bytes if needed.
+    while x > 0:
+        enc = enc + bchr(x & 0xFF)
+        x >>= 8
+
+    return enc
+
+
+def encode_str(x: bytes):
+    """Encode string function as defined in NIST SP 800-185"""
+
+    if x is None:
+        return left_encode(0)
+    else:
+        bitlen = len(x) * 8
+
+        if bitlen >= (1 << 2040):
+            raise ValueError("String too large to encode in cSHAKE128")
+
+        return left_encode(bitlen) + x
+
+
+def byte_align(x: bytes, length: int):
+    """Zero pad byte string as defined in NIST SP 800-185"""
+
+    x = left_encode(length) + x
+
+    # Note: this implementation works with byte aligned strings,
+    # hence no additional bit padding is needed at this point.
+    npad = (length - len(x) % length) % length
+    while npad:
+        x += bchr(0x00)
+        npad -= 1
+
+    return x
+
+
+class cSHAKE128_XOF(object):
+    """A cSHAKE128 hash object.
     Do not instantiate directly.
     Use the :func:`new` function.
-
-    :ivar oid: ASN.1 Object ID
-    :vartype oid: string
     """
 
-    # ASN.1 Object ID
-    oid = "2.16.840.1.101.3.4.2.11"
+    # Parameters
+    name = "cSHAKE128"
+    prefix_alignment = 168
+    keccak_capacity = c_size_t(32)
 
-    def __init__(self, data=None):
+    def __init__(self, data=None, function=None, custom=None):
         state = VoidPointer()
+
+        self.function = function
+        self.custom = custom
+
+        prefix = None
+        pad = 0x1F  # default to SHAKE
+        if function or custom:
+            prefix = encode_str(function) + encode_str(custom)
+            prefix = byte_align(prefix, self.prefix_alignment)
+            pad = 0x04  # for cSHAKE
```

Comment:
The the main functional changes are basically contained to these few lines, including lines 122 and 123 further below.